auth/gcloud/aio/auth/session.py

```diff
@@ -0,0 +1,157 @@
+    def delete(self, url: str, headers: Dict[str, str], params: Dict[str, str],
+               timeout: int):
+        pass
+
+    @abstractmethod
+    def request(self, method: str, url: str, headers: Dict[str, str],
+                auto_raise_for_status: bool = True, **kwargs: Any):
+        pass
+
+if not BUILD_GCLOUD_REST:
+    import aiohttp
+    class AioSession(BaseSession):
+        @property
+        def session(self) -> aiohttp.ClientSession:
+            self._session = self._session or aiohttp.ClientSession(
+                conn_timeout=self.conn_timeout, read_timeout=self.read_timeout)
+            return self._session
+
+        async def post(self, url: str, headers: Dict[str, str],
+                       data: str = None, timeout: int = 10,
+                       params: Dict[str, str] = None
+                       ) -> aiohttp.ClientResponse:
+            resp = await self.session.post(url, data=data, headers=headers,
+                                           timeout=timeout, params=params)
+            resp.raise_for_status()
+            return resp
+
+        async def get(self, url: str, headers: Dict[str, str] = None,
+                      timeout: int = 10, params: Dict[str, str] = None
+                      ) -> aiohttp.ClientResponse:
+            resp = await self.session.get(url, headers=headers, timeout=timeout,
+                                          params=params)
+            resp.raise_for_status()
+            return resp
+
+        async def put(self, url: str, headers: Dict[str, str], data: IOBase,
+                      timeout: int = 10) -> aiohttp.ClientResponse:
+            resp = await self.session.put(url, data=data, headers=headers,
+                                          timeout=timeout)
+            resp.raise_for_status()
+            return resp
+
+        async def delete(self, url: str, headers: Dict[str, str],
+                         params: Dict[str, str], timeout: int = 10
+                         ) -> aiohttp.ClientResponse:
+            resp = await self.session.delete(url, headers=headers,
+                                             params=params, timeout=timeout)
+            resp.raise_for_status()
+            return resp
+
+        async def request(self, method: str, url: str, headers: Dict[str, str],
+                          auto_raise_for_status: bool = True, **kwargs: Any
+                          ) -> aiohttp.ClientResponse:
+            resp = await self.session.request(method, url, headers=headers,
+                                              **kwargs)
+            if auto_raise_for_status:
+                resp.raise_for_status()
+            return resp
+
+if BUILD_GCLOUD_REST:
+    import requests
+
+    class SyncSession(BaseSession):
+        _google_api_lock = threading.RLock()
+
+        @property
+        def google_api_lock(self) -> threading.RLock:
+            return SyncSession._google_api_lock  # pylint: disable=protected-access
```

Comment:
I'm about 80% convinced this will let us maintain the locking semantics `gcloud-rest` needs. Sanity check: here's the case we need to cover:

```python
class Token(object):
    def __init__(self, ..., google_api_lock=None, ...):
        self.google_api_lock = google_api_lock or threading.RLock()

class KMS(object):
    def __init__(self, ..., google_api_lock=None, ...):
        self.google_api_lock = google_api_lock or threading.RLock()
        self.access_token = Token(..., google_api_lock=self.google_api_lock)

class Storage(object):
    def __init__(self, ..., google_api_lock=None, ...):
        self.google_api_lock = google_api_lock or threading.RLock()
        self.access_token = Token(..., google_api_lock=self.google_api_lock)
```

In the above example, if a user is making use of both the `KMS` and `Storage` classes, they need to be able to pass in a single `RLock` which is shared amongst *all requests made to Google*, ie. the lock used in `Token`, `KMS`, and `Storage` instances all need to be the same instance.

I _think_ what you've done here is ensure that users can't / don't need to pass in the lock, but we will use the same one everywhere anyway. Is that the case? And, further sanity check: is that the case in py2 as well as py3?

Reply:
> I think what you've done here is ensure that users can't / don't need to pass in the lock, but we will use the same one everywhere anyway. Is that the case?
Yes

> And, further sanity check: is that the case in py2 as well as py3?
Yes, the "static" class variable behaves the same way in both versions.

Reply:
Awesome, thanks for the sanity check!